# Praktikum 1
Praktikum 1 ini akan membuat JST sederhana (2 layer) dengan forward pass dan backpropagation manual. 

Backpropagation adalah sebuah algoritma untuk melatih jaringan saraf tiruan dengan cara mengoreksi kesalahan. Algoritma ini bekerja dengan cara menghitung selisih antara keluaran yang dihasilkan jaringan dan keluaran yang seharusnya (kesalahan), lalu memperbarui bobot dan bias jaringan secara berulang dari keluaran ke masukan untuk meminimalkan kesalahan tersebut. Cara kerja backpropagation

- Perambatan maju (forward pass): Data masukan diproses melalui jaringan dari lapisan masukan ke lapisan keluaran untuk menghasilkan prediksi awal.

- Hitung kesalahan: Selisih antara keluaran prediksi dan keluaran target dihitung menggunakan fungsi kerugian.

- Perambatan mundur (backward pass): Kesalahan disebarkan kembali ke belakang melalui jaringan dari lapisan keluaran ke lapisan masukan untuk menghitung gradien atau turunan parsial dari fungsi kerugian terhadap bobot dan bias.

- Perbarui bobot: Bobot dan bias disesuaikan menggunakan algoritma penurunan gradien untuk mengurangi kesalahan pada iterasi berikutnya. 

Langkah:

1. Buat dataset sederhana (XOR).
2. Inisialisasi bobot dan bias.
3. Implementasikan forward pass.
4. Hitung error dan lakukan backpropagation.
5. Update bobot menggunakan gradient descent.

In [4]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.2536515826608368
Epoch 1000, Loss: 0.2495729158496684
Epoch 2000, Loss: 0.24696569122085738
Epoch 3000, Loss: 0.22860406037052547
Epoch 4000, Loss: 0.1894753418120332
Epoch 5000, Loss: 0.15322680661168053
Epoch 6000, Loss: 0.047792811690261366
Epoch 7000, Loss: 0.021015782609133584
Epoch 8000, Loss: 0.01265723347331426
Epoch 9000, Loss: 0.008853101972329063
Prediksi:
[[0.07990118]
 [0.91015722]
 [0.91361817]
 [0.07088484]]


Tugas 1:

- Ubah jumlah neuron hidden layer menjadi 3.
- Bandingkan hasil loss dengan konfigurasi awal.
- Tambahkan fungsi aktivasi ReLU dan bandingkan hasil.


In [5]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3 
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.28964519741071004
Epoch 1000, Loss: 0.14255914507570708
Epoch 2000, Loss: 0.028147922016011444
Epoch 3000, Loss: 0.011122691196810294
Epoch 4000, Loss: 0.006475391200721984
Epoch 5000, Loss: 0.0044627358842315185
Epoch 6000, Loss: 0.003368258823032204
Epoch 7000, Loss: 0.0026889678509507777
Epoch 8000, Loss: 0.00222957482096295
Epoch 9000, Loss: 0.0018996641127322156
Prediksi:
[[0.02316956]
 [0.95931016]
 [0.95856373]
 [0.05195615]]


### Perbandingan Hasil Loss

Konfigurasi Awal (hidden_size = 2):
- Loss akhir (epoch 9000): 0.00885
- Prediksi: [[0.0799], [0.9102], [0.9136], [0.0709]]

Konfigurasi Baru (hidden_size = 3):
- Loss akhir (epoch 9000): 0.00190
- Prediksi: [[0.0232], [0.9593], [0.9586], [0.0520]]

Kesimpulan:
Hidden layer dengan 3 neuron menghasilkan loss yang lebih rendah (0.00190 vs 0.00885) dan prediksi yang lebih akurat mendekati target [0,1,1,0]. Penambahan neuron meningkatkan kapasitas model untuk mempelajari pola XOR.

In [18]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.37111477114413627
Epoch 1000, Loss: 0.25121420068636435
Epoch 2000, Loss: 0.25043446464997665
Epoch 3000, Loss: 0.25024791092528975
Epoch 4000, Loss: 0.25016878360425815
Epoch 5000, Loss: 0.2501260323538126
Epoch 6000, Loss: 0.2500995871358575
Epoch 7000, Loss: 0.2500817479080045
Epoch 8000, Loss: 0.2500689619224045
Epoch 9000, Loss: 0.2500593867642063
Prediksi:
[[1.20197835e-02]
 [6.92045740e-05]
 [9.88913008e-01]
 [8.88198586e-03]]
Epoch 9000, Loss: 0.2500593867642063
Prediksi:
[[1.20197835e-02]
 [6.92045740e-05]
 [9.88913008e-01]
 [8.88198586e-03]]


### Perbandingan Hasil Sebelum dan Sesudah Aktivasi ReLU

Sigmoid di Hidden Layer (hidden_size = 3):
- Loss akhir (epoch 9000): 0.00190
- Prediksi: [[0.0232], [0.9593], [0.9586], [0.0520]]

ReLU di Hidden Layer (hidden_size = 3):
- Loss akhir (epoch 9000): 0.2501
- Prediksi: [[0.0120], [0.0001], [0.9889], [0.0088]]

Kesimpulan:
Untuk masalah XOR, Sigmoid lebih baik daripada ReLU. ReLU mengalami kesulitan konvergensi (loss tinggi 0.2501 vs 0.00190) karena neuron bisa "mati" (dead neurons) ketika gradien menjadi 0 untuk nilai negatif. Sigmoid tetap memiliki gradien di semua titik, sehingga lebih cocok untuk jaringan shallow dan problem non-linear seperti XOR.